### ISSUE TITLE
aimet places QDQ nodes in between non‑computational nodes

### ISSUE THREAD

**Issue by yberkat** [quic/aimet#4074]
Hello AIMET team, I hope you're doing well ^^

I have a small question related to the ONNX QDQ export from torch models, Why does aimet insert QDQ nodes between non-computational (layout nodes) like Reshape/Transpose/Unsqueeze too ( they seem to have the same scale & zero_point of the preceding ops) ? Is this a design choice ? Is there a way to deactivate it ?

I'm sending a picture from a model i'm working on as an example !

<img width="1532" height="953" alt="Image" src="https://github.com/user-attachments/assets/ffaae141-14dd-4487-b45b-c74988dc04d9" />

Thank you in advance for your support ^^.

**Comment by quic-kyunggeu** [comment]
Hi @yberkat, thanks for filing this good question.

**Short answer** - Yes, it was a deliberate design choice that you can't deactivate.

**Long answer** - The reason we place QDQ to all nodes (including non-computational nodes like Tranpspose/Reshape/etc) is to explicitly indicate the data type of the tensor. Without those QDQ nodes, the target backend compiler might end up interpret the nodes without QDQ as floating points, which will result in significant performance drop due to extra dtype conversion/memcopy.

If you are concerned that those extra QDQs might cause performance drop, I'd like to clarify that they don't harm on-target performance; rather, removing them might cause performance drop.

**Comment by 2502912390** [comment]
Hi @quic-kyunggeu “Without those QDQ nodes, the target backend compiler might end up interpret the nodes without QDQ as floating points” I may need to clarify this point. While performing QAT with aimet_torch on a multi-input model, I observed that aimet_torch only quantizes one of the inputs. For the input that is not quantized, I found that QNN applies quantization to it during the qnn-onnx-converter stage.

At present, this behavior may be the reason why the inference results between aimet_torch and QNN cannot be aligned. Do you have any recommended solutions or best practices to address this issue?

<img width="1920" height="1014" alt="Image" src="https://github.com/user-attachments/assets/752ad97d-3d57-4d63-b2aa-c420f6f5b1c6" />

**Comment by quic-kyunggeu** [comment]
Looks like qnn-onnx-converter might have incorrectly inserted quantization encoding for `input_uv` that wasn't provided by AIMET.
To strictly align simulation accuracy and on-target accuracy, I recommend trying out [**qairt-converter**](https://docs.qualcomm.com/doc/80-63442-10/topic/converters.html#qairt-converter) with --use_quantize_v2 flag (beta feature). `qairt-converter` is the new converter that replaces `qnn-onnx-converter`. qairt-converter itself is stable; --use-quantize_v2 option is the beta feature

-----

More details for your understanding

#### 1. Why did AIMET not export encoding for `input_uv`?
The reason why `input_uv` has no encoding is twofold.
1. `input_uv` was fed into torch.cat function without getting quantized by any aimet [QuantizeDequantize](https://github.com/quic/aimet/blob/1973a07cd21d7981f697beaf5096b525bc3fb2a8/TrainingExtensions/torch/src/python/aimet_torch/v2/quantization/affine/quantizer.py#L884) module
2. `input_uv` has multiple consumers

Even though 1 is true, usually aimet would have reused Concat's output encoding as `input_uv`'s encoding **if** Concat was the only consumer of `input_uv`. (`concat(quantize(x), quantize(y))` and `quantize(concat(x, y))` are mathematically equivalent)
However, in your case there is a second consumer of `input_uv`. In this case, it's not mathematically equivalent to insert encoding for `input_uv` due to the second consumer.

#### 2. How does `qairt-converter --use_quantize_v2` (beta feature) help?
The reason you've got an encoding for input_uv under `qnn-onnx-converter` is probably one of the two.
* Common user mistakes, including
    * Passing --input_list argument
_-- This will calibrate the missing encodings based on the statistics of the inputs provided to --input_list_
    * Not passing --float_fallback option
_-- This option leaves the unquantized tensors as floating point instead of inserting artificial quantization encoding_
* Due to some bug in qnn-onnx-converter

On the other hand, `qairt-converter` is better than `qnn-onnx-converter` in both terms.
It prevents common user mistakes by better CLI design (--float_fallback is always ON, --input_list argument is removed), many bugs were fixed, and it's designed to preserve the accuracy of the input model especially with `--use_quantize_v2` option enabled. I definitely recommend you try out this one!

**Comment by 2502912390** [comment]
Hi @quic-kyunggeu, thank you very much for your suggestions. I will try using QAIRT-Converter as you recommended.

Regarding your first point:
You’re right that
   1. the current model’s input_uv is indeed passed into a torch.cat operation, and
   2. it is later added to an output via a residual connection (i.e., it has multiple consumers), which could explain why no quantization encodings are generated for this input.
Does this mean that AIMET does not support residual connections on model inputs?

<img width="420" height="503" alt="Image" src="https://github.com/user-attachments/assets/843ea396-22b7-48f5-8d8f-7ea41c64b120" />

However, in another model, this input
   1. is not passed through torch.cat, and
   2. does not have multiple consumers,yet it still does not get quantized.
Would you be able to explain why this might be happening?

<img width="1087" height="695" alt="Image" src="https://github.com/user-attachments/assets/c1ed7836-309a-454c-9ace-95c6dd158638" />
<img width="490" height="620" alt="Image" src="https://github.com/user-attachments/assets/dc2c5cd4-8bfb-47e4-8b9b-d1052b1b168a" />


Regarding your second point:
   1. When using qnn-onnx-converter, the --input_list argument is mandatory; otherwise, quantization will not be applied. Even if --quantization_overrides is provided, --input_list is still required.
   2. The option --float_fallback (which leaves unquantized tensors as floating point) is not applicable in the current input_y / input_uv model, because QNN does not allow concatenation of tensors with different data types. If this option were used, input_y would be quantized to int while input_uv would remain float, which is not allowed.

<img width="1022" height="1002" alt="Image" src="https://github.com/user-attachments/assets/54c14ef9-e1b6-4463-9968-388f62446b4c" />

At this point, I am more inclined to believe that AIMET-Torch does compute the quantization parameters during QAT/PTQ, but that these parameters are lost during the export stage due to certain limitations or issues. Otherwise, the observed behavior does not seem to be logically consistent.

**Comment by yberkat** [comment]
> 
> If you are concerned that those extra QDQs might cause performance drop, I'd like to clarify that they don't harm on-target performance; rather, removing them might cause performance drop.

Hello @quic-kyunggeu, thank you for addressing these concerns, I just found it weird at first, because when using an older version of  `aimet_torch` in the past; QDQ nodes were inserted only between computational nodes (I might be wrong ofc), is this a recent change ?

**NB**: I'm using `aimet_torch` to perform PTQ and `aimet_torch.onnx.export` to generate the ONNX.

Also; I'm also facing a similar issue with a multi-inputs model (inputs + RNN states), where the RNN states don't have QDQ nodes (encodings) attached to them (They have multipe consumers, so that might be the reason for it) as opposed to the model input..., I sent a picture below to illustrate this !

- RNN states
<img width="621" height="785" alt="Image" src="https://github.com/user-attachments/assets/c5a5edec-5b97-4a27-a4f2-5af0511c8da2" />

- Model input
<img width="621" height="785" alt="Image" src="https://github.com/user-attachments/assets/95f05d3e-0420-4fca-83df-3411d93b8e8d" />

**Comment by quic-kyunggeu** [comment]
@yberkat 
> I just found it weird at first, because when using an older version of aimet_torch in the past; QDQ nodes were inserted only between computational nodes (I might be wrong ofc), is this a recent change ?

Yes, it was  first added in aimet 2.18 (17883b913), which is relatively still new.


> Also; I'm also facing a similar issue with a multi-inputs model (inputs + RNN states), where the RNN states don't have QDQ nodes (encodings) attached to them (They have multipe consumers, so that might be the reason for it) as opposed to the model input..., I sent a picture below to illustrate this !

I'm not exactly sure why GRU doesn't have input QDQ. This is most likely because the quantsim never created input quantizer for those, but I failed to reproduce your problem.

I used this toy script, and the hidden state input was followed by QDQ nodes when exported to onnx. Can you try this and see if you can reproduce this error with minimal changes? I tested my script with torch 2.10 & aimet 2.20-2.24.
```py
import torch
import aimet_torch


class Model(torch.nn.Module):
    def __init__(self):
        super(Model, self).__init__()
        self.gru = torch.nn.GRU(input_size=10, hidden_size=20, num_layers=2)

    def forward(self, *args):
        output, _ = self.gru(*args)
        return output


def main():
    model = Model()
    dummy_input = (
        torch.randn(5, 3, 10),  # (seq_len, batch_size, input_size)
        torch.randn(2, 3, 20),  # (num_layers, batch_size, hidden_size)
    )
    sim = aimet_torch.QuantizationSimModel(model, dummy_input[0])
    sim.compute_encodings(lambda model: model(*dummy_input))

    aimet_torch.onnx.export(
        sim.model,
        dummy_input,
        "gru.onnx",
        dynamo=False,
        input_names=["input", "state_in_0"],
        output_names=["output"],
    )



if __name__ == "__main__":
    main()
```

**Comment by quic-kyunggeu** [comment]
@2502912390 
> Does this mean that AIMET does not support residual connections on model inputs?

**Short answer:** Not exactly. If the input is fed into multiple nn.Modules, the input will be quantized by each module's input quantizer(s).

**Long answer**: AIMET quantizes the model by converting all nn.Module into corresponding Quantized- modules (such as [QuantizedConv2d](https://github.com/quic/aimet/blob/9972c1b5bfabcfa1b1487fecbe2aee16e94ca591/TrainingExtensions/torch/src/python/aimet_torch/v2/nn/true_quant.py#L922)). This means the inputs & outputs of standalone functional operators like torch.cat, torch.add, etc. will **NOT** be quantized by aimet usually. This is the reason you had no input quantizer in the first case - because it was going into functional operator `torch.cat`.

> does not have multiple consumers,yet it still does not get quantized.
Would you be able to explain why this might be happening?

I think it's probably either because nn.Linear's input quantizer was never created or was deleted afterwards. Are you using a custom Linear module (not torch.nn.Linaer) by any chance? Personally I've never seen this case before.


> 1. When using qnn-onnx-converter, the --input_list argument is mandatory; otherwise, quantization will not be applied. Even if --quantization_overrides is provided, --input_list is still required.
> 2. The option --float_fallback (which leaves unquantized tensors as floating point) is not applicable in the current input_y / input_uv model, because QNN does not allow concatenation of tensors with different data types. If this option were used, input_y would be quantized to int while input_uv would remain float, which is not allowed.

I think you might have misunderstood my comment.
In my original comment I advised **not to use qnn-onnx-converter**. Try qairt-converter instead. Both issues must have been already resolved there

> At this point, I am more inclined to believe that AIMET-Torch does compute the quantization parameters during QAT/PTQ, but that these parameters are lost during the export stage due to certain limitations or issues. Otherwise, the observed behavior does not seem to be logically consistent.

I can give you a firm answer about this. If a quantizer exists and it was used during `aimet_torch.onnx.export`, it will be always exported to onnx. Rather, a more likely situation is that the quantizer may not have been created at all in the first place, which is the only case you won't see in the exported onnx QDQ. (See the long answer in my first paragraph)

**Comment by yberkat** [comment]
Hello @quic-kyunggeu, thank you for answering my previous questions ^^

To follow up with the second one, I'm sending an example code to help illustrate the lack of QDQ nodes on the  `states_in` (it happens when we have two sequential GRUs and states are split and concatenated inside the forward function).

```python
import torch
from torch import nn
import aimet_torch
from aimet_torch.v2.quantsim import QuantizationSimModel

class MultiInputModel(nn.Module):
    def __init__(self):
        super().__init__()
        # Model parameters
        self.freq_bins = 257 
        self.embedding_dim = 128
        self.hidden_channels = 32
        self.gru_hidden_size = 64
        self.gru_input_size = self.hidden_channels * self.freq_bins
        
        # 1. First Conv layer
        self.conv1 = nn.Conv2d(in_channels=1, 
                               out_channels=self.hidden_channels, 
                               kernel_size=(3, 3), 
                               padding=(1, 1))
        
        # 1.5 Two sequential GRU layers to model temporal dependencies
        # Flattening Freq and Channels for GRU input: hidden_channels * freq_bins
        self.gru1 = nn.GRU(input_size=self.gru_input_size, 
                           hidden_size=self.gru_hidden_size, 
                           batch_first=True)
                           
        # Second GRU takes output of first (same size)
        self.gru2 = nn.GRU(input_size=self.gru_hidden_size, 
                           hidden_size=self.gru_hidden_size, 
                           batch_first=True)
        
        # Projection back from GRU -> Conv
        self.gru_proj = nn.Linear(self.gru_hidden_size, self.gru_input_size)
        
        # 3. Second Conv layer (output)
        self.conv2 = nn.Conv2d(in_channels=self.hidden_channels, 
                               out_channels=1, 
                               kernel_size=(3, 3), 
                               padding=(1, 1))
        
        self.relu = nn.ReLU()

    def forward(self, x_spec, h_in):
        # x_spec: [B, 1, F, T]
        # h_in: [2, B, H] (Stacked GRU hidden states for 2 layers)
        
        # 1. Conv branch
        out_conv = self.relu(self.conv1(x_spec)) # [B, C, F, T]
        batch, channels, freqs, time = out_conv.shape
        
        # 2. GRU Processing
        # Permute to [B, T, C, F] -> flatten to [B, T, C*F]
        gru_in = out_conv.permute(0, 3, 1, 2).reshape(batch, time, -1)
        
        # Separate hidden states
        # h_in is [2, B, H], split into two [1, B, H] tensors
        h1_in, h2_in = torch.split(h_in, 1, dim=0)
        
        # Run GRU 1
        # Use contiguous check or clone if needed, but split creates view usually safe for inputs
        gru1_out, h1_out = self.gru1(gru_in, h1_in) # [B, T, H], [1, B, H]
        
        # Run GRU 2 (takes output of GRU 1)
        gru2_out, h2_out = self.gru2(gru1_out, h2_in) # [B, T, H], [1, B, H]
        
        # Concatenate output states
        h_out = torch.cat([h1_out, h2_out], dim=0) # [2, B, H]
        
        # Project back from last GRU output
        gru_in_proj = self.gru_proj(gru2_out) # [B, T, C*F]
        
        # Reshape back to [B, C, F, T]
        # First reshape to [B, T, C, F] then permute to [B, C, F, T]
        out_rnn = gru_in_proj.reshape(batch, time, channels, freqs).permute(0, 2, 3, 1)

        # 4. Combine (Skip connection from Conv + RNN Path)
        combined = out_conv + out_rnn
        
        # 5. Output Conv
        out = self.conv2(combined)
        
        return out, h_out

def calibration_pass(model, args):
    # args is expected to be the dummy input tuple
    x_spec, h_in = args
    # Run a few forward passes
    model.eval()
    with torch.no_grad():
        for _ in range(1):
            model(x_spec, h_in)

def main():    
    # Initializing the model and dummy inputs
    model = MultiInputModel()
    # Batch=1, Channels=1, Freq=257, Time=100
    dummy_input_spec = torch.randn(1, 1, 257, 100)
    # GRU state: [num_layers=2, batch=1, hidden_size=64]
    dummy_h_in = torch.randn(2, 1, 64)
    dummy_input = (dummy_input_spec, dummy_h_in)
    
    # Init QuantizationSimModel
    sim = QuantizationSimModel(model, dummy_input=dummy_input)
    
    # Calibration
    sim.compute_encodings(forward_pass_callback=calibration_pass, 
                          forward_pass_callback_args=dummy_input)
    
    aimet_torch.onnx.export(
        sim.model,
        dummy_input,
        "simple_se2.onnx",
        dynamo=False,
        input_names=["x_spec", "h_in"],
        output_names=["output", "h_out"],
    )

if __name__ == "__main__":
    main()
```

However, when using a single GRU layer, I notice that we have QDQ nodes on `state_in` and `state_out` of the GRU (as illustrated in the picture below).

<img width="893" height="569" alt="Image" src="https://github.com/user-attachments/assets/0d9be2c8-d248-4da8-9ff4-cc513360971b" />

Thank you again for your support ^^.

**NB:** I'm using `torch==2.8` & `aimet_torch==2.24`

**Comment by quic-kyunggeu** [comment]
@yberkat I believe your last question was answered in https://github.com/quic/aimet/issues/4078. Closing this issue
